File: code-recipes/php/src/easyship/GetHandoverSlotsRecipe.php
<?php

namespace Src\easyship;

use SpApi\Api\easyship\v2022_03_23\EasyShipApi;
use SpApi\Model\easyship\v2022_03_23\Dimensions;
use SpApi\Model\easyship\v2022_03_23\ListHandoverSlotsRequest;
use SpApi\Model\easyship\v2022_03_23\UnitOfLength;
use SpApi\Model\easyship\v2022_03_23\UnitOfWeight;
use SpApi\Model\easyship\v2022_03_23\Weight;
use Src\util\Recipe;

/**
 * Code Recipe to get available handover time slots for EasyShip orders
 * Steps:
 * 1. Setup order and package details
 * 2. Initialize EasyShip API client
 * 3. Create request with package dimensions and weight
 * 4. Call listHandoverSlots API
 * 5. Display available time slots
 */
class GetHandoverSlotsRecipe extends Recipe
{
    private EasyShipApi $easyShipApi;
    private string $amazonOrderId;
    private string $marketplaceId;

    public function start(): void
    {
        $this->setupOrderDetails();
        $this->initializeEasyShipApi();
        $response = $this->listHandoverSlots();
        $this->displayTimeSlots($response);
        echo "✅ Successfully retrieved handover slots\n";
    }

    private function setupOrderDetails(): void
    {
        $this->amazonOrderId = "702-3035602-4225066";
        $this->marketplaceId = "A1AM78C64UM0Y8";
        echo "Order details configured: {$this->amazonOrderId}\n";
    }

    private function initializeEasyShipApi(): void
    {
        $this->easyShipApi = new EasyShipApi($this->config);
        echo "EasyShip API client initialized\n";
    }

    private function listHandoverSlots()
    {
        $request = new ListHandoverSlotsRequest();
        $request->setAmazonOrderId($this->amazonOrderId);
        $request->setMarketplaceId($this->marketplaceId);
        $request->setPackageDimensions($this->createPackageDimensions());
        $request->setPackageWeight($this->createPackageWeight());

        $response = $this->easyShipApi->listHandoverSlots($request);
        echo "Handover slots retrieved for order: {$this->amazonOrderId}\n";
        return $response;
    }

    private function createPackageDimensions(): Dimensions
    {
        // TODO: In production, retrieve actual dimensions from CalculateOrderDimensionsRecipe
        // These are sample values for demonstration purposes only
        $dimensions = new Dimensions();
        $dimensions->setLength(10.0);
        $dimensions->setWidth(8.0);
        $dimensions->setHeight(5.0);
        $dimensions->setUnit(UnitOfLength::CM);
        return $dimensions;
    }

    private function createPackageWeight(): Weight
    {
        // TODO: In production, retrieve actual dimensions from CalculateOrderDimensionsRecipe
        // These are sample values for demonstration purposes only
        $weight = new Weight();
        $weight->setValue(500.0);
        $weight->setUnit(UnitOfWeight::G);
        return $weight;
    }

    private function displayTimeSlots($response): void
    {
        $timeSlots = $response->getTimeSlots();
        if (!$timeSlots || empty($timeSlots)) {
            echo "No time slots available\n";
            return;
        }
        
        echo "Available time slots: " . count($timeSlots) . "\n";
        foreach ($timeSlots as $slot) {
            echo "  Slot ID: " . $slot->getSlotId() . "\n";
            echo "    Method: " . $slot->getHandoverMethod() . "\n";
        }
    }
}

File: code-recipes/java/src/main/java/easyship/GetHandoverSlotsRecipe.java
package easyship;

import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.easyship.v2022_03_23.EasyShipApi;
import software.amazon.spapi.models.easyship.v2022_03_23.Dimensions;
import software.amazon.spapi.models.easyship.v2022_03_23.ListHandoverSlotsRequest;
import software.amazon.spapi.models.easyship.v2022_03_23.ListHandoverSlotsResponse;
import software.amazon.spapi.models.easyship.v2022_03_23.TimeSlot;
import software.amazon.spapi.models.easyship.v2022_03_23.UnitOfLength;
import software.amazon.spapi.models.easyship.v2022_03_23.UnitOfWeight;
import software.amazon.spapi.models.easyship.v2022_03_23.Weight;
import util.Constants;
import util.Recipe;

import com.amazon.SellingPartnerAPIAA.LWAException;

/**
 * Code Recipe to get available handover time slots for EasyShip orders
 * Steps:
 * 1. Setup order and package details
 * 2. Initialize EasyShip API client
 * 3. Create request with package dimensions and weight
 * 4. Call listHandoverSlots API
 * 5. Display available time slots
 */
public class GetHandoverSlotsRecipe extends Recipe {

    private EasyShipApi easyShipApi;
    private String amazonOrderId;
    private String marketplaceId;

    @Override
    protected void start() {
        setupOrderDetails();
        initializeEasyShipApi();
        ListHandoverSlotsResponse response = listHandoverSlots();
        displayTimeSlots(response);
        System.out.println("✅ Successfully retrieved handover slots");
    }

    private void setupOrderDetails() {
        amazonOrderId = "702-3035602-4225066";
        marketplaceId = "A1AM78C64UM0Y8";
        System.out.println("Order details configured: " + amazonOrderId);
    }

    private void initializeEasyShipApi() {
        easyShipApi = new EasyShipApi.Builder()
                .lwaAuthorizationCredentials(lwaCredentials)
                .endpoint(Constants.BACKEND_URL)
                .disableAccessTokenCache()
                .build();
        System.out.println("EasyShip API client initialized");
    }

    private ListHandoverSlotsResponse listHandoverSlots() {
        try {
            ListHandoverSlotsRequest request = new ListHandoverSlotsRequest()
                    .amazonOrderId(amazonOrderId)
                    .marketplaceId(marketplaceId)
                    .packageDimensions(createPackageDimensions())
                    .packageWeight(createPackageWeight());

            ListHandoverSlotsResponse response = easyShipApi.listHandoverSlots(request);
            System.out.println("Handover slots retrieved for order: " + amazonOrderId);
            return response;
        } catch (ApiException | LWAException e) {
            throw new RuntimeException("Failed to list handover slots", e);
        }
    }

    private Dimensions createPackageDimensions() {
        // TODO: In production, retrieve actual dimensions from CalculateOrderDimensionsRecipe
        // These are sample values for demonstration purposes only
        return new Dimensions()
                .length(10.0f)
                .width(8.0f)
                .height(5.0f)
                .unit(UnitOfLength.CM);
    }

    private Weight createPackageWeight() {
        // TODO: In production, retrieve actual dimensions from CalculateOrderDimensionsRecipe
        // These are sample values for demonstration purposes only
        return new Weight()
                .value(500.0f)
                .unit(UnitOfWeight.G);
    }

    private void displayTimeSlots(ListHandoverSlotsResponse response) {
        if (response.getTimeSlots() == null || response.getTimeSlots().isEmpty()) {
            System.out.println("No time slots available");
            return;
        }
        System.out.println("Available time slots: " + response.getTimeSlots().size());
        for (TimeSlot slot : response.getTimeSlots()) {
            System.out.println("  Slot ID: " + slot.getSlotId());
            System.out.println("    Method: " + slot.getHandoverMethod());
        }
    }
}
